server/src/main/java/org/elasticsearch/transport/RemoteClusterAwareClient.java

```diff
@@ -37,6 +45,16 @@
         Request request,
         ActionListener<Response> listener
     ) {
+        if (ensureConnected == false) {
+            try {
+                remoteClusterService.getConnection(clusterAlias);
+            } catch (NoSuchRemoteClusterException e) {
+                listener.onFailure(e);
+                // trigger another connection attempt, but don't wait for it to complete
+                remoteClusterService.ensureConnected(clusterAlias, ActionListener.wrap(() -> {}));
+                return;
+            }
+        }
         remoteClusterService.ensureConnected(clusterAlias, ActionListener.wrap(v -> {
```

Comment:
This still risks a wait in cases where we discovered the remote cluster was disconnected just after calling `getConnection()` above. I think we should avoid waiting on the result of `ensureConnected()` at all if `this.ensureConnected` is false.

Reply:
If the connection is blackholed (and still appears healthy), we might also wait on that. I'm not sure if it's worth complicating the implementation here for this edge case. This is a best effort check for cases where connections are unavailable for a prolonged time, not a perfect failure detector.

Reply:
The trouble I see is that we can handle an active-but-blackholed connection pretty quickly (a few seconds of retries is enough) but the wait for a new connection is always much slower. With a high enough rate of searches we could end up with many of them blocked within `ensureConnected`. I was thinking of something like the following, are you sure this is too complicated?

```java
    @Override
    protected <Request extends ActionRequest, Response extends ActionResponse> void doExecute(
        ActionType<Response> action,
        Request request,
        ActionListener<Response> listener
    ) {
        maybeEnsureConnected(ActionListener.wrap(v -> {
            final Transport.Connection connection;
            try {
                if (request instanceof RemoteClusterAwareRequest) {
                    DiscoveryNode preferredTargetNode = ((RemoteClusterAwareRequest) request).getPreferredTargetNode();
                    connection = remoteClusterService.getConnection(preferredTargetNode, clusterAlias);
                } else {
                    connection = remoteClusterService.getConnection(clusterAlias);
                }
            } catch (NoSuchRemoteClusterException e) {
                if (ensureConnected == false) {
                    // trigger another connection attempt, but don't wait for it to complete
                    remoteClusterService.ensureConnected(clusterAlias, ActionListener.wrap(() -> {}));
                }
                throw e;
            }
            service.sendRequest(
                connection,
                action.name(),
                request,
                TransportRequestOptions.EMPTY,
                new ActionListenerResponseHandler<>(listener, action.getResponseReader())
            );
        }, listener::onFailure));
    }

    private void maybeEnsureConnected(ActionListener<Void> ensureConnectedListener) {
        if (ensureConnected) {
            remoteClusterService.ensureConnected(clusterAlias, ensureConnectedListener);
        } else {
            ensureConnectedListener.onResponse(null);
        }
    }
```

Reply:
++ looks good to me. I've pushed the change